### DISCUSSION TITLE
Limited utility of NPM package for contract projects

### DISCUSSION THREAD

**Discussion by andreivladbrg** [sablier-labs/evm-monorepo#601] [General]
One of the main motivation for launching our project on npm was to simplify the installation process of `v2-core` for integrators using the Hardhat environment.

However, due to the [remappings](https://github.com/sablier-labs/v2-core/blob/main/remappings.txt) we've configured, this becomes impossible without installing extra plugins.

For instance, if you're looking to incorporate the contracts into your Hardhat project via `pnpm install @sablier/v2-core`, you would import and use them as follows:
```solidity
import "hardhat/console.sol";
import "@sablier/v2-core/src/SablierV2LockupLinear.sol";

contract Example {
    SablierV2LockupLinear internal sablier;

    constructor(SablierV2LockupLinear _sablier) {
        sablier = _sablier;
    }

    function foo() public {
        console.log("The next stream id is: ", sablier.nextStreamId());
    }
}

```

The code from above would revert due to [HH404](https://hardhat.org/hardhat-runner/docs/errors#HH404);
![image](https://github.com/sablier-labs/v2-core/assets/99738872/0343923e-bf4f-4ded-9d07-44fee3c1ad5f)

This is because we're importing prb math from the following path:

https://github.com/sablier-labs/v2-core/blob/e6b705be300bbbb5824911152768390031bd2175/src/SablierV2LockupLinear.sol#L7

But the correct path for hardhat should be `@prb/math/src/UD60x18.sol";`

Given Hardhat's lack of native support for remappings, such as:
```
@sablier/v2-core/=node_modules/@sablier/v2-core/src/
@openzeppelin/contracts/=node_modules/@openzeppelin/contracts/
@prb/math/=node_modules/@prb/math/src/
```
There are two solutions for this: 
1. [hardhat-preprocessor ](https://book.getfoundry.sh/config/hardhat#instructions) - a more complex route
2. [hardhat-foundry](https://hardhat.org/hardhat-runner/docs/advanced/hardhat-and-foundry) - a simpler approach and one recommended by Hardhat team 

This problem manifests the same way in Truffle. However, I think it wouldn't be an issue for [Brownie](https://eth-brownie.readthedocs.io/en/stable/compile.html#compile-remap), which supports remapping.

Given this, I'm left wondering: aside from facilitating front-end integration via React, are there any other reasons to maintain the npm package?

**Comment by PaulRBerg** [comment]
Thanks for investigating this problem, @andreivladbrg!

React integrations aren't the only reason for maintaining the npm package (although this alone would have been important!). An npm package is useful as a generic way to access the contract artifacts since the NPM registry is so ubiquitous.

For instance, imagine we are sponsoring a hackathon and talking to hackers building on top of Sablier. The hackers ask us to provide them with the contract artifacts. Thanks to the npm package deployment, we can share the following URL with them:

https://unpkg.com/browse/@sablier/v2-core/

Now, regarding the problem with Hardhat itself. The [`hardhat-foundry`](https://hardhat.org/hardhat-runner/docs/advanced/hardhat-and-foundry) solution is the way to go; that's what we will recommend to our users, too.

**Comment by PaulRBerg** [comment]
Closing - this issue is solvable by using the `hardhat-foundry` package.
